CONTRIBUTING.md

```diff
@@ -0,0 +1,233 @@
+
+  When platform specific code is required, it should be placed in `GVFSPlatform` or one of the platforms it contains (e.g. `IKernelDriver`)
+
+## Tracing and Logging
+
+- *The "Error" logging level is reserved for non-retryable errors that result in I/O failures or the VFS4G process shutting down*
+
+  The expectation from our customers is that when VFS4G logs an "Error" level message in its log file either:
+    * VFS4G had to shut down unexpectedly
+    * VFS4G encountered an issue severe enough that user-initiated I/O would fail.
+
+- *Log full exception stacks*
+
+  Full exception stacks (i.e. `Exception.ToString`) provide more details than the exception message alone (`Exception.Message`). Full exception stacks make root-causing issues easier.
+  
+- *Do not display full exception stacks to users*
+
+  Exception call stacks are not usually actionable for the user.  Users frequently (sometimes incorrectly) assume that VFS4G has crashed when shown a full stack.  The full stack *should* be included in VFS4G logs, but *should not* be displayed as part of the error message provided to the user.
+
+- *Include relevant details when logging exceptions*
+
+  Sometimes an exception call stack alone is not enough to root cause failures in VFS4G.  When catching (or throwing) exceptions, log relevant details that will help diagnose the issue.  As a general rule, the closer an exception is caught to where it's thrown, the more relevant details there will be to log.
+
+  Example:
+  ```
+  catch (Exception e)
+  {
+    EventMetadata metadata = new EventMetadata();
+    metadata.Add("Area", "Mount");
+    metadata.Add(nameof(enlistmentHookPath), enlistmentHookPath);
+    metadata.Add(nameof(installedHookPath), installedHookPath);
+    metadata.Add("Exception", e.ToString());
+    context.Tracer.RelatedError(metadata, $"Failed to compare {hook.Name} version");
+  }
+  ```
+
+## Error Handling
+
+- *Fail fast: An error or exception that risks data loss or corruption should shut down VFS4G immediately*
+
+  Preventing data loss and repository corruption is critical.  If an error or exception occurs that could lead to data loss, it's better to shut down VFS4G than keep the repository mounted and risk corruption.
+
+- *Do not catch exceptions that are indicative of a programmer error (e.g. `ArgumentNullException`)*
+
+  Any exceptions that result from programmer error (e.g. `ArgumentNullException`) should be discovered as early in the development process as possible.  Avoid `catch` statements that would hide these errors (e.g. `catch(Exception)`).
+
+  The only exception to this rule is for [unhandled exceptions in background threads](#bgexceptions)
+
+- *Do not use exceptions for normal control flow*
+
+  Prefer to write code that avoids exceptions being thrown (e.g. the `TryXXX` pattern).  There are performance costs to using exceptions for control flow, and VFS4G code typically needs to handle errors where (and as soon as) they occur.
```

Comment:
```VFS4G code typically needs to handle errors where (and as soon as) they occur.```

I am not sure what this means in this context...

Reply:
I did not word that well.

What I was trying to get across is that for many I/O errors VFS4G cares about exactly where the error occurred in a sequence of operations, and that determining that failure point be a bit clunky with exceptions.

I'll work on a better description for what I mean, I've included an example below, let me know if you have any suggestions on how to re-word things.

_Example with TryXXX_

```
bool TryDoWorkOnDisk(string fileContents, out string error)
{
  if (!TryCreateReadConfig())
  {
    error = "Failed to read config file";
    return false;
  }

  if (!TryCreateTempFile(fileContents))
  {
    error = "Failed to create temp file";
    return false;
  }

  if (!TryRenameTempFile())
  {
    error = "Failed to rename temp file";
    if (!TryDeleteTempFile())
    {
        error += ", and failed to cleanup temp file";
    }

    return false;
  }

  error = null;
  return true;
}
```

_Example 2: Same behavior using exceptions_

```
bool TryDoWorkOnDisk(string fileContents, out string error)
{
  try
  {
    CreateReadConfig();
  }
  catch (Exception ex) when (ex is IOException || ex is UnauthorizedAccessException)
  {
    error = "Failed to read config file";
    return false;
  }

  try
  {
    CreateTempFile(fileContents);
  }
  catch (Exception ex) when (ex is IOException || ex is UnauthorizedAccessException)
  {
    error = "Failed to create temp file";
    return false;
  }

  try
  {
    RenameTempFile();
  }
  catch (Exception ex) when (ex is IOException || ex is UnauthorizedAccessException)
  {
    error = "Failed to rename temp file";

    try
    {
      DeleteTempFile();
    }
    catch (Exception ex) when (ex is IOException || ex is UnauthorizedAccessException)
    {
        error += ", and failed to cleanup temp file";
    }

    return false;
  }

  error = null;
  return true;
}
```

_Example 3: Catch needs to figure out what went wrong_

```
bool TryDoWorkOnDisk(string fileContents, out string error)
{
  try
  {
    CreateReadConfig();
    CreateTempFile(fileContents);
    RenameTempFile();
  }
  catch (Exception ex) when (ex is IOException || ex is UnauthorizedAccessException)
  {
    error = "Something went wrong doing work on disk";

    try
    {
       if (TempFileExists())
       {
           DeleteTempFile();
        }
    }
    catch (Exception e) when (e is IOException || e is UnauthorizedAccessException)
    {
        error += ", and failed to cleanup temp file";
    }

    return false;
  }

  error = null;
  return true;
}
```